ambassador/kubewatch.py

```diff
@@ -331,25 +333,29 @@ def watch_loop(restarter):
     v1 = kube_v1()
 
     if v1:
-        w = watch.Watch()
-
-        if "AMBASSADOR_SINGLE_NAMESPACE" in os.environ:
-            watched = w.stream(v1.list_namespaced_service, namespace=restarter.namespace)
-        else:
-            watched = w.stream(v1.list_service_for_all_namespaces)
-
-        for evt in watched:
-            logger.debug("Event: %s %s/%s" % 
-                         (evt["type"], 
-                          evt["object"].metadata.namespace, evt["object"].metadata.name))
-            sys.stdout.flush()
-
-            if evt["type"] == "DELETED":
-                restarter.delete(evt["object"])
-            else:
-                restarter.update_from_service(evt["object"])
-
-        logger.info("watch loop exited?")
+        while True:
+            try:
+                w = watch.Watch()
+
+                if "AMBASSADOR_SINGLE_NAMESPACE" in os.environ:
+                    watched = w.stream(v1.list_namespaced_service, namespace=restarter.namespace)
+                else:
+                    watched = w.stream(v1.list_service_for_all_namespaces)
+
+                for evt in watched:
+                    logger.debug("Event: %s %s/%s" % 
+                                 (evt["type"], 
+                                  evt["object"].metadata.namespace, evt["object"].metadata.name))
+                    sys.stdout.flush()
+
+                    if evt["type"] == "DELETED":
+                        restarter.delete(evt["object"])
+                    else:
+                        restarter.update_from_service(evt["object"])
+
+                logger.info("watch loop exited?")
+            except ProtocolError:
+                continue
```

Comment:
Is it worth logging anything here?

Reply:
Sure. When the connection was lost and caught it, container should tell us that ambassador rewatched k8s automatically. I'll rewrite this to log a message as debug, and resend pull request. Thank you.